lib/fs/encoder.go

```diff
@@ -0,0 +1,326 @@
+// Copyright (C) 2021 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package fs
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/syncthing/syncthing/lib/sync"
+)
+
+const (
+	// The standard encoder is effectively how Syncthing has worked
+	// until now: It assumed all filenames could be saved without
+	// encoding, and that encoded filenames would not be found on disk,
+	// as there was no need to encode them (for example, ext4
+	// filesystems).
+	//
+	// So now, if the standard encoder finds encoded filenames, it ignores
+	// them, as they are assumed to have been written in error
+	// (see footnote*), as the underlying filesystem (say ext4) does not
+	// require filenames to be encoded to be saved.
+	//
+	// This differs from how the passthru encoder works, as it reads and
+	// writes filenames without encoding them beforehand.
+	//
+	// But this only works if all the other devices are configured to use
+	// the passthru encoder. This is because the standard encoder ignores
+	// encoded filenames, and the other encoders decode encoded
+	// filenames, so devices using the passthru encoder can only exchange
+	// encoded filenames with devices configured to use the passthru
+	// encoder.
+	//
+	// * For example, if the user selects the fat encoder, saves
+	// some encoded filenames, and then changes the encoder back to
+	// standard. On ext4 filesystems, this could lead to filenames being
+	// written both with their original names (using the standard
+	// encoder), and their encoded names (using the fat encoder).
+	//
+	// That's why the standard encoder must ignore encoded filenames when
+	// found on disk, to avoid "seeing" both the original filename, and
+	// its encoded equivalent.
```

Comment:
Does this mean that file names with the magic unicode characters in the, which are synced using `passthrough` or in previous versions of Syncthing, will be deleted when switching to `standard`?

Reply:
No, they wouldn't be deleted, but `standard` would now ignore them. To sync them, the user could use the `fat` encoder to sync them using their encoded names, or the `passthru` encoder to sync them as their encoded names.

Reply:
Thanks for reviewing. I think you're right, the `passthrough` encoder needs to be the default. I need to rethink this whole concept.